clarify CPU and memory limit enforcement differences (#46222)
kubernetes/website: content/en/docs/concepts/configuration/manage-resources-containers.md

The behavior of `cpu` and `memory` limit enforcement is slightly different.

`cpu` limits are enforced by CPU throttling. When a container approaches
its `cpu` limit, the kernel will restrict access to the CPU corresponding to the
container's limit. Thus, a `cpu` limit is a hard limit the kernel enforces.
Containers may not use more CPU than is specified in their `cpu` limit.

`memory` limits are enforced by the kernel with out of memory (OOM) kills. When
a container uses more than its `memory` limit, the kernel may terminate it. However,
terminations only happen when the kernel detects memory pressure. Thus, a
container that over allocates memory may not be immediately killed. This means
`memory` limits are enforced reactively. A container may use more memory than
its `memory` limit, but if it does, it may get killed.

{{< note >}}
There is an alpha feature `MemoryQoS` which attempts to add more preemptive
limit enforcement for memory (as opposed to reactive enforcement by the OOM
killer). However, this effort is
[stalled](https://github.com/kubernetes/enhancements/tree/a47155b340/keps/sig-node/2570-memory-qos#latest-update-stalled)

> due to a potential livelock situation a memory hungry can cause.

Reviewer: Did you mean: a memory hungry **application**?

a Pod scheduled to a Node with 8GiB of memory and no other Pods, then the container can try to use
more RAM.

> Both `CPU` and `memory` limits are applied by the kubelet (and
> {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}),
> and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
> by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `CPU` and `memory` limits

Reviewer 2: ```suggestion
Both `cpu` and `memory` limits are applied by the kubelet (and
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}),
and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `cpu` and `memory` limits
```

> Both `CPU` and `memory` limits are applied by the kubelet (and
> {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}),
> and are ultimately enforced by the kernel. On Linux nodes, limits are enforced
> by {{< glossary_tooltip text="cgroups" term_id="cgroup" >}}). However, `CPU` and `memory` limits
> are enforced differently.

Reviewer 2: > However, `CPU` and `memory` limits
 are enforced differently.

The above statement is wrong. You already answered correctly in lines 37 - 38, yet contradicted your answer in lines 40 - 41. There's no need to throw the kernel into the conversation here, it ends up confusing the reader.

Only two things are missing from the existing documentation.
1. An outline that the kubelet and container runtime [communicates](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#cgroup-drivers) with **control groups** to enforce the limit.
2. A configured `memory.high` setting in cgroupv2 can prevent a container from getting killed by an OOM error when it exceeds its limit.

Since these are what you are trying to add, I suggest you use the existing documentation and only include what is missing instead of changing all the sentences.

```suggestion
If you set a `memory` limit of 4GiB for that container, the kubelet (and
 {{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}) communicates with control groups to enforce the limit.
 The runtime prevents the container from using more than the configured resource limit. For example:
 when a process in the container tries to consume more than the allowed amount of memory,
 the system kernel terminates the process that attempted the allocation, with an out of memory
```

Author: I have clarified this section, I think it's important to mention who is actually doing the enforcing and how. I did mention MemoryQoS but I don't think we should go in depth because we're not moving forward with that effort now.

Reviewer 2: Nice!

> 

Reviewer 2: I'll suggest you leave this as it is.
```suggestion
(OOM) error.
```

Author: this line was dropped because the paragraph was reworked, so I think we should keep it dropped

Reviewer 2: Alright

> `CPU` limits are enforced by CPU throttling. When a container approaches

Reviewer 3: Can we link a follow-up resource on how CPU throttling works in linux?

Reviewer 2: I removed the follow-up resource. I don't think changing this line is necessary. See this [review](https://github.com/kubernetes/website/pull/46222/files#r1793695009).

Author: > Can we link a follow-up resource on how CPU throttling works in linux?

I can't find a good one that isn't on some companies website. the cgroup documentation doesn't really go into depth on how it works. Any suggestions?

Reviewer 2: I previously used this one from [Wikipedia](https://en.wikipedia.org/wiki/Dynamic_frequency_scaling) since its open source.

Author: hm maybe I'm using throttling wrong in this case. That article looks like it's describing behavior of the hardware of scaling its density up and down to match workloads. In this case, this is really CPU time slicing, where a process gets slices of the CPU time according to its weight, and is limited in going above that CPU time limit

> `CPU` limits are enforced by CPU throttling. When a container approaches

Reviewer 2: ```suggestion
`cpu` limits are enforced by [CPU throttling](https://en.wikipedia.org/wiki/Dynamic_frequency_scaling) When a container approaches
```

> `CPU` limits are enforced by CPU throttling. When a container approaches

Reviewer 2: ```suggestion
`cpu` limits are enforced by [CPU throttling](https://en.wikipedia.org/wiki/Dynamic_frequency_scaling) When a container approaches
```

> its `CPU` limit, the kernel will restrict access to the CPU corresponding to the

Reviewer 2: ```suggestion
its `cpu` limit, the kernel will restrict access to the CPU corresponding to the
```

> container's limit.

Reviewer: `cpu` (lower case) please @haircommander; the original is wrong, but we should fix it. In the actual API, it's `cpu`.

Author: I tried updating so whenever the limit is specified I used lowercase

> `CPU` limits are enforced by CPU throttling. When a container approaches
> its `CPU` limit, the kernel will restrict access to the CPU corresponding to the
> container's limit.

Reviewer 2: This too.
I'll suggest you reference a resource showing how kubernetes implements CPU throttling, a concept similar to the `memory.high` settings in [cgroupv2](https://kubernetes.io/blog/2023/05/05/qos-memory-resources/#how-it-works).
```suggestion
Limits can be implemented either reactively (the system intervenes once it sees a violation)
or by enforcement (the system prevents the container from ever exceeding the limit). Different
 runtimes can have different ways to implement the same restrictions.
```

> `memory` limits are enforced by the kernel with OOM (out of memory) kills. When

Reviewer 3: ```suggestion
`memory` limits are enforced by the kernel with out of memory (OOM) kills. When
```

Author: good point!

a pod uses more than its `memory` limit, the kernel may terminate it. However,
if there is available memory on the node, the kernel may choose not to terminate

> the offending pod.

Reviewer: 1. What happens to memory allocation attempts -  `sbrk()` and friends - that would assign memory above the limit, if they were to succeed? Ideally, let's be clear about where the enforcement does and doesn't happen.

1. The container runtime or kernel might apply a higher limit than you request. For example, if you specify a limit of 2097159 bytes (a shade over 2MiB), the kernel is likely to round things up to a whole page size. That's if your container runtime even lets you request a limit that small!

Reviewer 4: I don't agree to this change related to memory limits. It is not accurate and it will lead to more confusion, because memory limit enforcement is a pretty complex topic.

Reviewer 2: Is it because the kubelet or container runtime uses a [cgroup driver](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#cgroup-drivers) to interact with the OS’s underlying cgroup system to enforce a limit, hence, any limit created by a different cgroup does not matter?

Reviewer 4: @network-charles Did you read this? https://kubernetes.io/blog/2023/05/05/qos-memory-resources/

Reviewer 2: I just did. It's similar to my last comment but with an added info on the “memory.high” feature supported by cgroupv2. It's clear now.

We may need to clarify that the kernel uses the “memory. high” feature supported on cgroupv2 to throttle the container. Also, the statement that it applies to a pod is wrong, according to the blog, it only applies at the container cgroups level.

Reviewer 4: So ... the point is ... a Pod may and may not get killed when it is using more memory than its limit.
Can we rephrase this to be more accurate?

Author: I added a note about memoryQOS feature, it's not something we should document heavily because the effort is stalled in SIG node due to issues

Author: > What happens to memory allocation attempts - sbrk() and friends - that would assign memory above the limit, if they were to succeed? Ideally, let's be clear about where the enforcement does and doesn't happen.

it's unfortunately not so easy. When the kernel enforces OOM kills, it usually is in response to a memory allocation. so sometimes, sbrk and friends will trigger an oom kill of the process, and other times it will succeed.

> The container runtime or kernel might apply a higher limit than you request. For example, if you specify a limit of 2097159 bytes (a shade over 2MiB), the kernel is likely to round things up to a whole page size. That's if your container runtime even lets you request a limit that small!

I don't know if we need to get to this level of detail here, but I'm open to suggestions

> 
> `memory` limits are enforced by the kernel with OOM (out of memory) kills. When
> a pod uses more than its `memory` limit, the kernel may terminate it. However,
> if there is available memory on the node, the kernel may choose not to terminate
> the offending pod.

Reviewer 2: ```suggestion
In [cgroup v2](https://kubernetes.io/docs/concepts/architecture/cgroups/), if you configure the `memory.high` [settings](https://docs.kernel.org/admin-guide/cgroup-v2.html#memory) and a
pod uses more than its memory limit, the kernel will throttle the memory of the pod.
However, this throttling will only occur if the node has enough available memory.
```

Author: we stalled on this effort so i don't think we should mention it deeply
